File: graphedu/common/models/vo/systemv2/function.py
"""功能权限管理相关 VO 模型 (View Objects - 响应模型)

职责：
1. 定义 API 响应的数据结构
2. 配置序列化规则（from_attributes=True 支持从 ORM 对象创建）
"""

from datetime import datetime

from pydantic import Field

from graphedu.common.models.vo.base import VO


class FunctionListVO(VO):
    """功能列表项 VO"""

    function_id: int = Field(description="功能ID")
    parent_id: int = Field(description="父功能ID")
    function_name: str = Field(description="功能名称")
    function_key: str | None = Field(default=None, description="权限标识（GROUP/DIVIDER 类型为 None）")
    function_type: str = Field(
        description="功能类型: DIR-目录, MENU-菜单, BUTTON-按钮, INTERFACE-接口, GROUP-菜单分组, DIVIDER-菜单分隔线"
    )
    route_path: str | None = Field(default=None, description="路由路径")
    route_cache: str | None = Field(
        default=None, description="路由路径页面是否缓存（Y是 N否，对应sys_data_option字典）"
    )
    route_query: dict | None = Field(default=None, description="路由传递参数")
    route_external: str | None = Field(default=None, description="是否外链（Y是 N否，对应sys_data_option字典）")
    component: str | None = Field(default=None, description="组件路径")
    layout_component: str | None = Field(default=None, description="布局组件路径")
    icon: str | None = Field(default=None, description="图标")
    sort_order: int = Field(description="显示顺序")
    visible: str = Field(description="是否可见（Y是 N否，对应sys_data_option字典）")
    style: dict | None = Field(default=None, description="菜单CSS样式（JSON格式）")
    option_style: dict | None = Field(default=None, description="菜单选项样式（JSON格式）")
    status: str = Field(description="状态（0正常 1停用）")
    scene: str = Field(default="admin", description="应用场景: web-日常应用, admin-管理系统, mobile-移动端")
    create_time: datetime | None = Field(default=None, description="创建时间")


class FunctionDetailVO(VO):
    """功能详细信息 VO"""

    function_id: int = Field(description="功能ID")
    parent_id: int = Field(description="父功能ID")
    function_name: str = Field(description="功能名称")
    function_key: str | None = Field(default=None, description="权限标识（GROUP/DIVIDER 类型为 None）")
    function_type: str = Field(
        description="功能类型: DIR-目录, MENU-菜单, BUTTON-按钮, INTERFACE-接口, GROUP-菜单分组, DIVIDER-菜单分隔线"
    )
    route_path: str | None = Field(default=None, description="路由路径")
    route_cache: str | None = Field(default=None, description="路由路径页面是否缓存（Y是 N否）；仅 MENU 类型有效")
    route_query: dict | None = Field(default=None, description="路由传递参数")
    route_external: str | None = Field(default=None, description="是否外链（Y是 N否）；仅 MENU 类型有效")
    component: str | None = Field(default=None, description="组件路径")
    layout_component: str | None = Field(default=None, description="布局组件路径（如: layout/CommonLayout/index）")
    icon: str | None = Field(default=None, description="图标")
    sort_order: int = Field(description="显示顺序")
    visible: str = Field(description="是否可见（Y是 N否，对应sys_data_option字典）")
    style: dict | None = Field(default=None, description="菜单CSS样式（JSON格式，使用css-in-js格式）")
    option_style: dict | None = Field(default=None, description="菜单选项样式（JSON格式）")
    status: str = Field(description="对照sys_data_status（0正常 1停用 2已删除）")
    scene: str = Field(default="admin", description="应用场景: web-日常应用, admin-管理系统, mobile-移动端")
    create_by: int | None = Field(default=None, description="创建者")
    create_time: datetime | None = Field(default=None, description="创建时间")
    update_by: int | None = Field(default=None, description="更新者")
    update_time: datetime | None = Field(default=None, description="更新时间")
    remark: str | None = Field(default=None, description="备注")


class FunctionTreeBriefVO(VO):
    """功能树节点 简要 VO（用于树形选择展示）"""

    function_id: int = Field(description="功能ID")
    parent_id: int = Field(description="父功能ID")
    function_name: str = Field(description="功能名称")
    function_type: str = Field(
        description="功能类型: DIR-目录, MENU-菜单, BUTTON-按钮, INTERFACE-接口, GROUP-菜单分组, DIVIDER-菜单分隔线"
    )
    children: list["FunctionTreeBriefVO"] = Field(default_factory=list, description="子功能列表")


class FunctionTreeVO(VO):
    """功能树节点 VO（用于最外层页面树形展示）"""

    function_id: int = Field(description="功能ID")
    parent_id: int = Field(description="父功能ID")
    function_name: str = Field(description="功能名称")
    function_key: str | None = Field(default=None, description="权限标识（GROUP/DIVIDER 类型为 None）")
    function_type: str = Field(
        description="功能类型: DIR-目录, MENU-菜单, BUTTON-按钮, INTERFACE-接口, GROUP-菜单分组, DIVIDER-菜单分隔线"
    )
    route_path: str | None = Field(default=None, description="路由路径")
    route_cache: str | None = Field(default=None, description="路由页面是否缓存")
    route_external: str | None = Field(default=None, description="是否外链")
    component: str | None = Field(default=None, description="组件路径")
    layout_component: str | None = Field(default=None, description="布局组件路径（如: layout/CommonLayout/index）")
    icon: str | None = Field(default=None, description="图标")
    sort_order: int = Field(description="显示顺序")
    visible: str = Field(description="是否可见")
    status: str = Field(description="状态")
    scene: str = Field(default="admin", description="应用场景: web-日常应用, admin-管理系统, mobile-移动端")
    create_time: datetime = Field(default=None, description="创建时间")
    has_children: bool | None = Field(default=None, description="是否有子功能")
    children: list["FunctionTreeVO"] | None = Field(default_factory=list, description="子功能列表")


class RoleFunctionTreeVO(VO):
    """角色功能树 VO（用于分配权限时展示）"""

    checked_ids: list[int] = Field(default_factory=list, description="已分配的功能ID列表")
    function_trees: list[FunctionTreeVO] = Field(default_factory=list, description="功能树列表")

File: graphedu-ui/src/types/api/system/function.ts
/**
 * 功能权限管理相关类型定义
 * 对应后端：graphedu/common/models/dto/function.py 和 vo/function.py
 */

// ============================================================================
// 请求 DTO 类型
// ============================================================================

/**
 * 功能类型枚举
 */
export type FunctionType = 'DIR' | 'MENU' | 'BUTTON' | 'INTERFACE' | 'GROUP' | 'DIVIDER'

/**
 * 功能查询请求
 */
export interface FunctionQueryDTO {
  /** 功能名称（模糊查询） */
  functionName?: string
  /** 功能状态，对照 sys_data_status（0正常 1停用） */
  status?: '0' | '1'
  /** 是否可见（Y是 N否，对应 sys_data_option 字典） */
  visible?: 'Y' | 'N'
  /** 功能类型: DIR-目录, MENU-菜单, BUTTON-按钮, INTERFACE-接口, GROUP-分组, DIVIDER-分隔线 */
  functionType?: FunctionType
  /** 应用场景: web-日常应用, admin-管理系统, mobile-移动端 */
  scene?: string
}

/**
 * 创建功能请求
 */
export interface FunctionCreateDTO {
  /** 父功能ID（0表示根节点） */
  parentId?: number
  /** 功能名称 */
  functionName: string
  /** 权限标识（如: student:list, course:add）；GROUP/DIVIDER 类型无需填写 */
  functionKey?: string
  /** 功能类型: DIR-目录, MENU-菜单, BUTTON-按钮, INTERFACE-接口, GROUP-分组, DIVIDER-分隔线 */
  functionType: FunctionType
  /** 路由路径 */
  routePath?: string
  /** 路由页面是否缓存（N不缓存 Y缓存）；仅 MENU 类型有效 */
  routeCache?: 'N' | 'Y'
  /** 路由传递参数（JSON格式） */
  routeQuery?: Record<string, any>
  /** 是否外链（N否 Y是）；仅 MENU 类型有效 */
  routeExternal?: 'N' | 'Y'
  /** 组件路径 */
  component?: string
  /** 布局组件路径（如: layout/CommonLayout/index, layout/WorkbenchLayout/index） */
  layoutComponent?: string
  /** 菜单图标 */
  icon?: string
  /** 显示顺序 */
  sortOrder?: number
  /** 是否可见（N隐藏 Y显示） */
  visible?: 'N' | 'Y'
  /** 菜单CSS样式（JSON格式，使用css-in-js格式） */
  style?: Record<string, any>
  /** 菜单选项样式（JSON格式） */
  optionStyle?: Record<string, any>
  /** 功能状态，对照 sys_data_status（0正常 1停用） */
  status?: '0' | '1'
  /** 应用场景: web-日常应用, admin-管理系统, mobile-移动端 */
  scene?: string
  /** 备注 */
  remark?: string
}

/**
 * 更新功能请求
 */
export interface FunctionUpdateDTO {
  /** 功能ID */
  functionId: number
  /** 父功能ID */
  parentId?: number
  /** 功能名称 */
  functionName?: string
  /** 权限标识 */
  functionKey?: string
  /** 功能类型 */
  functionType?: FunctionType
  /** 路由路径 */
  routePath?: string
  /** 路由页面是否缓存（N不缓存 Y缓存）；仅 MENU 类型有效 */
  routeCache?: 'N' | 'Y'
  /** 路由传递参数 */
  routeQuery?: Record<string, any>
  /** 是否外链（N否 Y是）；仅 MENU 类型有效 */
  routeExternal?: 'N' | 'Y'
  /** 组件路径 */
  component?: string
  /** 布局组件路径（如: layout/CommonLayout/index, layout/WorkbenchLayout/index） */
  layoutComponent?: string
  /** 菜单图标 */
  icon?: string
  /** 显示顺序 */
  sortOrder?: number
  /** 是否可见（N隐藏 Y显示） */
  visible?: 'N' | 'Y'
  /** 菜单CSS样式（JSON格式，使用css-in-js格式） */
  style?: Record<string, any>
  /** 菜单选项样式（JSON格式） */
  optionStyle?: Record<string, any>
  /** 功能状态，对照 sys_data_status（0正常 1停用） */
  status?: '0' | '1'
  /** 应用场景: web-日常应用, admin-管理系统, mobile-移动端 */
  scene?: string
  /** 备注 */
  remark?: string
}

// ============================================================================
// 响应 VO 类型
// ============================================================================

/**
 * 功能树节点
 */
export interface FunctionTreeVO {
  /** 功能ID */
  functionId: number
  /** 父功能ID */
  parentId: number
  /** 功能名称 */
  functionName: string
  /** 权限标识；GROUP/DIVIDER 类型为 undefined */
  functionKey?: string
  /** 功能类型，对照 sys_function_type（DIR目录, MENU菜单, BUTTON按钮, INTERFACE接口, GROUP菜单分组, DIVIDER菜单分隔线） */
  functionType: FunctionType
  /** 路由路径 */
  routePath?: string
  /** 路由路径页面是否缓存，对应 sys_data_option（Y是 N否）；仅 MENU 类型有效 */
  routeCache?: string
  /** 是否外链，对应 sys_data_option（Y是 N否）；仅 MENU 类型有效 */
  routeExternal?: string
  /** 组件路径 */
  component?: string
  /** 布局组件路径（如: layout/CommonLayout/index） */
  layoutComponent?: string
  /** 图标 */
  icon?: string
  /** 菜单CSS样式（JSON格式，使用css-in-js格式） */
  style?: Record<string, any>
  /** 菜单选项样式（JSON格式） */
  optionStyle?: Record<string, any>
  /** 路由传递参数（JSON格式） */
  routeQuery?: Record<string, any>
  /** 显示顺序 */
  sortOrder: number
  /** 是否可见，对应 sys_data_option（Y是 N否） */
  visible: string
  /** 功能状态，对照 sys_data_status（0正常 1停用） */
  status: string
  /** 应用场景，对照 sys_function_scene（web日常应用 admin管理系统 userInfo个人中心） */
  scene: string
  /** 创建时间 */
  createTime?: string
  /** 是否有子功能 */
  hasChildren?: boolean
  /** 子功能列表 */
  children?: FunctionTreeVO[]
}

/**
 * 功能树节点（简要版，用于下拉选择）
 */
export interface FunctionTreeBriefVO {
  /** 功能ID */
  functionId: number
  /** 父功能ID */
  parentId: number
  /** 功能名称 */
  functionName: string
  /** 功能类型: DIR-目录, MENU-菜单, BUTTON-按钮, INTERFACE-接口, GROUP-分组, DIVIDER-分隔线 */
  functionType: FunctionType
  /** 子功能列表 */
  children?: FunctionTreeBriefVO[]
}

/**
 * 功能详细信息
 */
export interface FunctionDetailVO {
  /** 功能ID */
  functionId: number
  /** 父功能ID */
  parentId: number
  /** 功能名称 */
  functionName: string
  /** 权限标识；GROUP/DIVIDER 类型为 undefined */
  functionKey?: string
  /** 功能类型: DIR-目录, MENU-菜单, BUTTON-按钮, INTERFACE-接口, GROUP-分组, DIVIDER-分隔线 */
  functionType: FunctionType
  /** 路由路径 */
  routePath?: string
  /** 路由页面是否缓存（N不缓存 Y缓存）；仅 MENU 类型有效 */
  routeCache?: string
  /** 路由传递参数 */
  routeQuery?: Record<string, any>
  /** 是否外链（N否 Y是）；仅 MENU 类型有效 */
  routeExternal?: string
  /** 组件路径 */
  component?: string
  /** 布局组件路径（如: layout/CommonLayout/index） */
  layoutComponent?: string
  /** 图标 */
  icon?: string
  /** 菜单CSS样式（JSON格式，使用css-in-js格式） */
  style?: Record<string, any>
  /** 菜单选项样式（JSON格式） */
  optionStyle?: Record<string, any>
  /** 显示顺序 */
  sortOrder: number
  /** 是否可见（0隐藏 1显示） */
  visible: string
  /** 功能状态，对照 sys_data_status（0正常 1停用） */
  status: string
  /** 应用场景: web-日常应用, admin-管理系统, mobile-移动端 */
  scene: string
  /** 创建者 */
  createBy?: number
  /** 创建时间 */
  createTime?: string
  /** 更新者 */
  updateBy?: number
  /** 更新时间 */
  updateTime?: string
  /** 备注 */
  remark?: string
}

/**
 * 角色功能树（用于分配权限时展示）
 */
export interface RoleFunctionTreeVO {
  /** 已分配的功能ID列表 */
  checkedIds: number[]
  /** 功能树列表 */
  functionTrees: FunctionTreeVO[]
}
